app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+        } catch (ex) {
+          console.error(v2RulesetPath + ' stream invalid regexp ' + data.key + ': ' + ex.toString())
+        }
+      }
+
+      let result
+      try {
+        result = JSON.parse(data.value)
+      } catch (ex) {
+        console.error(v2RulesetPath + ' stream invalid JSON ' + data.entry + ': ' + data.value)
+      }
+
+      done(null, result.exclude)
+    }).on('error', (err) => {
+      console.error(v2RulesetPath + ' stream error: ' + JSON.stringify(err, null, 2))
+    }).on('close', () => {
+    }).on('end', () => {
+      if (!doneP) done(null, false)
+    })
+  })
+}
+
+const setLocation = (state, timestamp, tabId) => {
+  if (!synopsis) {
+    return state
+  }
+
+  const locationData = ledgerState.getLocation(state, currentUrl)
+  if (_internal.verboseP) {
+    console.log(
+      `locations[${currentUrl}]=${JSON.stringify(locationData, null, 2)} ` +
+      `duration=${(timestamp - currentTimestamp)} msec tabId= ${tabId}`
+    )
+  }
+  if (locationData.isEmpty() || !tabId) {
+    return state
+  }
+
+  let publisherKey = locationData.get('publisher')
+  if (!publisherKey) {
+    return state
+  }
+
+  if (!visitsByPublisher[publisherKey]) {
+    visitsByPublisher[publisherKey] = {}
+  }
+
+  if (!visitsByPublisher[publisherKey][currentUrl]) {
+    visitsByPublisher[publisherKey][currentUrl] = {
+      tabIds: []
+    }
+  }
+
+  const revisitP = visitsByPublisher[publisherKey][currentUrl].tabIds.indexOf(tabId) !== -1
+  if (!revisitP) {
+    visitsByPublisher[publisherKey][currentUrl].tabIds.push(tabId)
+  }
+
+  let duration = timestamp - currentTimestamp
+  if (_internal.verboseP) {
+    console.log('\nadd publisher ' + publisherKey + ': ' + duration + ' msec' + ' revisitP=' + revisitP + ' state=' +
+      JSON.stringify(underscore.extend({location: currentUrl}, visitsByPublisher[publisherKey][currentUrl]),
+        null, 2))
+  }
+
+  synopsis.addPublisher(publisherKey, {duration: duration, revisitP: revisitP})
+  state = ledgerState.setPublisher(state, publisherKey, synopsis.publishers[publisherKey])
+  state = updatePublisherInfo(state)
+  state = verifiedP(state, publisherKey, (error, result) => {
+    if (!error) {
+      appActions.onPublisherOptionUpdate(publisherKey, 'verified', result)
+      savePublisherOption(publisherKey, 'verified', result)
+    }
+  })
+
+  return state
+}
+
+const addVisit = (state, location, timestamp, tabId) => {
+  if (location === currentUrl) {
+    return state
+  }
+
+  state = setLocation(state, timestamp, tabId)
+
+  const lastUrl = pageDataState.getLastUrl(state)
+  const aboutUrl = getSourceAboutUrl(lastUrl) || lastUrl
+  if (aboutUrl.match(/^about/)) {
+    state = pageDataState.resetInfo(state)
+  }
+
+  location = getSourceAboutUrl(location) || location
+
+  currentUrl = location.match(/^about/) ? locationDefault : location
+  currentTimestamp = timestamp
+  return state
+}
+
+const getFavIcon = (state, publisherKey, page) => {
```

Comment:
these could benefit from a falsey check... something like:
```
const protocol = page && page.get('protocol')
if (protocol && publisher && !publisher.get('protocol')) {
```

Reply:
you don't need this check, because `getPublisher` always return Immutable.Map

```
getPublisher: (state, key) => {
    state = validateState(state)
    if (key == null) {
      return Immutable.Map()
    }

    return state.getIn(['ledger', 'synopsis', 'publishers', key]) || Immutable.Map()
  }
```